Libs/RaptureCore/src/main/java/rapture/script/reflex/ReflexRaptureScript.java

```diff
@@ -287,39 +288,100 @@ public ScriptResult runProgramExtended(CallingContext context, IActivityInfo act
-                ReflexValue v = walker.currentScope.resolve(val.getKey());
-                if (v != null && v.getValue() != ReflexValue.Internal.VOID && v.getValue() != ReflexValue.Internal.NULL) {
-                    log.debug("Injecting " + v.asObject() + " as " + val.getKey());
-                    val.setValue(v.asObject());
-                } else walker.currentScope.assign(val.getKey(), val.getValue() == null ? new ReflexNullValue() : new ReflexValue(val.getValue()));
-
+        ReflexTreeWalker walker = getParserWithStandardContext(context, script.getScript(), params);
+        ProgressDebugger progress = (timeout > 0) ? new TimeoutReflexDebugger(activity, script.getScript(), timeout)
+                : new ProgressDebugger(activity, script.getScript());
+        // Setup an alternate output handler, and a standard data handler
+        walker.getReflexHandler().setDataHandler(new ReflexDataHelper(context));
+        walker.getReflexHandler().setOutputHandler(new ScriptResultOutputHandler(res));
+        ReflexNode execRes = walker.walk();
+
+        ReflexExecutor.injectSystemIntoScope(walker.currentScope);
+
+        MetaScriptInfo scriptInfo = walker.getScriptInfo();
+
+        for (Map.Entry<String, Object> entry : params.entrySet()) {
+            String key = entry.getKey();
+            String value = entry.getValue().toString();
+            ReflexValue v = walker.currentScope.resolve(key);
+            MetaParam param = scriptInfo.getParameter(key);
+            if (param != null) {
+                switch (param.getParameterType()) {
+                case STRING:
+                default:
+                    v = new ReflexValue(value);
+                    break;
+                case VOID:
+                    v = new ReflexVoidValue(-1);
+                    break;
+                case NUMBER:
+                    v = new ReflexValue(new BigDecimal(value.toString()));
+                    break;
+                case INTEGER:
+                    v = new ReflexValue(new BigDecimal(value.toString()).intValue());
+                    break;
+                case BOOLEAN:
+                    v = new ReflexValue(Boolean.valueOf(value.toString()));
+                    break;
+                case MAP:
+                    v = new ReflexValue(JacksonUtil.getMapFromJson(value.toString()));
+                    break;
+                case LIST:
+                    List<ReflexValue> list = JacksonUtil.objectFromJson(value.toString(), ArrayList.class);
+                    v = new ReflexValue(list);
+                    break;
+                }
             }
-            // TODO replace this with abortable invocation
-            if (timeout > 0) log.info("Warning: script is not abortable");
-            ReflexValue val = execRes.evaluateWithoutScope(progress);
-            progress.getInstrumenter().log();
+            walker.currentScope.assign(key, (v == null) ? new ReflexNullValue() : v);
```

Comment:
Ah, the test for null is unnecessary. 
scriptInfo.getParameter(key) always returns a value. 
If none is defined in the meta block it creates one of type String. 
I'll change that to prevent confusion.